File: internal/web/i18n.go
package web

import (
	"encoding/json"
	"net/http"
	"strings"

	assets "github.com/AnaTofuZ/freekiosk-manager/web"
)

var english = loadLanguage("en")
var japanese = loadLanguage("ja")

func loadLanguage(lang string) map[string]string {
	data, err := assets.Files.ReadFile("locales/" + lang + ".json")
	if err != nil {
		panic(err)
	}
	var text map[string]string
	if err := json.Unmarshal(data, &text); err != nil {
		panic(err)
	}
	return text
}

func language(r *http.Request) string {
	if lang := r.URL.Query().Get("lang"); lang == "en" || lang == "ja" {
		return lang
	}
	if cookie, err := r.Cookie("language"); err == nil && cookie.Value == "ja" {
		return "ja"
	}
	return "en"
}

func translations(lang string) map[string]string {
	if lang == "ja" {
		return japanese
	}
	return english
}

func translate(lang, message string) string {
	if lang == "ja" {
		for key, value := range english {
			if value == message {
				return japanese[key]
			}
		}
		if strings.HasPrefix(message, "Device returned HTTP ") {
			return "HTTPエラー (" + strings.TrimPrefix(message, "Device returned HTTP ") + ")"
		}
	}
	return message
}

File: web/src/i18n.ts
import japanese from "../locales/ja.json";
import english from "../locales/en.json";

export const locale = document.documentElement.lang;

export function translate(message: string): string {
  if (locale !== "ja") return message;
  const text: Record<string, string> = japanese;
  const key = Object.entries(english).find(([, value]) => value === message)?.[0];
  if (key) return text[key];
  const http = /^(?:Device returned HTTP |Request failed \(HTTP )(\d{3})\)?$/.exec(message);
  if (http) return `HTTPエラー (${http[1]})`;
  return message;
}
